desktop-exporter/app/utils/array-to-tree.ts

```diff
@@ -67,9 +68,74 @@ export function arrayToTree(spans: SpanData[]): TreeItem[] {
     }
   }
 
-  // In order to handle incomplete traces, the missing spans get appended to the end of the rootItems array
-  // This way we make sure that the root span is added first (if present), followed by any missing spans
-  for (let spanID of missingSpanIDs) {
+  // To handle incomplete traces:
+  // 1. Sort the missing spans by the earliest start time of their children
+  // 2. Appended sorted spand to the end of the rootItems array
+  // This way we make sure that the root span is added first (if present),
+  // and preserve the visual clarity of the waterfall view
+  let missingSpans = Array.from(missingSpanIDs).sort((a: string, b: string) => {
```

Comment:
One option to simplify this logic is more precise types. Something like:

```typescript
export type RootTreeItem =
  | {
      status: SpanDataStatus.present;
      spanData: SpanData;
    children: TreeItem[];
    }
  | {
      status: SpanDataStatus.missing;
      spanID: string;
    children: TreeItem[];
    };

export type TreeItem = {
  spanData: SpanData;
  children: TreeItem[];
}
```

So that missing spans can only happen in the root of an `RootTreeItem[]`

Baring that, a helper function could help here:

```typescript
function getEarliestStartTime(children: TreeItem[]): number {
  if (children.length == 0) {
    throw new Error(
      "Unexpected type: A 'missing' parent span appears to have no children.",
    );
  }

  let startTimes = children.map((treeItem) => {
    if (treeItem.status === SpanDataStatus.missing) {
      throw new Error(
        "Unexpected type: A child of a 'missing' parent span appears to have no SpanData.",
      );
    }
    // getNsFromString returns `any` instead of `number`
    // this could also be fixed
    return getNsFromString(treeItem.spanData.startTime) as number;
  });

  return Math.min(...startTimes)
}
```

Reply:
TFW you forget Math.min exists. It's much nicer now!